core/src/main/java/org/springframework/security/authentication/MFATokenEvaluator.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.authentication;
+
+import org.springframework.security.core.Authentication;
+
+/**
+ * Evaluates <code>Authentication</code> tokens
+ *
+ * @author Yoshikazu Nojima
+ */
+public interface MFATokenEvaluator {
+
+	/**
+	 * Indicates whether the passed <code>Authentication</code> token represents a
+	 * user in the middle of multi factor authentication process.
+     *
+	 * @param authentication to test (may be <code>null</code> in which case the method
+	 * will always return <code>false</code>)
+	 *
+	 * @return <code>true</code> the passed authentication token represented a principal
+	 * in the middle of multi factor authentication process, <code>false</code> otherwise
+	 */
+	boolean isMultiFactorAuthentication(Authentication authentication);
+
+	/**
+	 * Indicates whether the principal associated with the <code>Authentication</code>
+	 * token is allowed to login with only single factor.
+	 *
+	 * @param authentication to test (may be <code>null</code> in which case the method
+	 * will always return <code>false</code>)
+	 *
+	 * @return <code>true</code> the principal associated with thepassed authentication
+	 * token is allowed to login with only single factor, <code>false</code> otherwise
+	 */
+	boolean isSingleFactorAuthenticationAllowed(Authentication authentication);
```

Comment:
The APIs need to consider that multi factor authentication is really an authorization requirement. It isn't sufficient to be able to answer if single factor authentication allowed. There needs to be the ability to answer:

Does any specific part of the application require another authentication factor? If so, we need to be able to request that specific factor from the user. There might be different types of factors that are triggered at different places within the application. A page that uses a credit card might require that the user has submitted their password again recently, a user with settings enabled for MFA may require the user to enter a code sent over SMS to access any part of the application, an administrator may require FIDO2 to access administrative parts of the applications, etc.

A possibility is that we should set these rules using granted authorities.

Reply:
The objective of this API is to check the user is allowed to login with single authentication factor or not.
It is intended to be used in MultiFactorAuthenticationProvider.
https://github.com/spring-projects/spring-security/pull/6842/files#diff-d87a3d8131f824e9c9a173b2a907dc92R68

To realize access-controll based on authentication level, yes, we should use granted authority.
I suppose users can implement it by handling the `AuthenticationSuccessEvent`, but it is not the scope of `isSingleFactorAuthenticationAllowed` method.